### PR TITLE
Fix imporing of non-OpenSSH Ed25519 keys

### PR DESCRIPTION
**Pull request by oakkitten**
This is a rare and elusive key, and the reason why no-one has ever ran into this issue is because ConnectBot seems to be the only way to create such keys and use them with sshd. Oh you can absolutely create a key like that with `openssl`, however, `ssh-keygen` doesn't seem to be able to read such keys, private or public, in any way, so you can't easily produce the .pub format sshd can use.

When sshlib can't itself import a key, we are trying with Bouncy Castle. For the older RSA etc keys it all works without a hitch as these implement well-known Java interfaces. EdDsa keys are newer and have separate interfaces in Bouncy Castle and sshlib. Sshlib can't import Non-OpenSSH Ed25519 keys itself, so what we end up with a Bouncy Castle-specific key, which is, while very much a valid key, not usable by sshlib. To mitigate that, this adds conversion from BC-specific Ed25519 key to its sshlib-specific variant.

Also, this fixes a bug in our code that created Ed25519 public keys from Ed25519 private keys in a wrong way.

### CODE REVIEW

File: app/src/main/java/androidx/preference/PrivateKeyPickerPreference.kt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+@Suppress("ACCIDENTAL_OVERRIDE")
+class PrivateKeyPickerPreference(context: Context?, attrs: AttributeSet?) :
+        PasswordedFilePickerPreference(context, attrs) {
+
+    // First, try decoding the key pair using our SSH library.
+    // If that fails due to unsupported key formats, try decoding the key pair using BC.
+    //
+    // Then, check if we have a pair of EdDSAKey. It is a valid key pair,
+    // but since sshlib is not friendly with BC, it can't use it.
+    // Assume it's a pair of Ed25519 keys, and covert it to sshlib format.
```
Comment (zopieux) [inline app/src/main/java/androidx/preference/PrivateKeyPickerPreference.kt:36]: "convert"

File: app/src/main/java/androidx/preference/PrivateKeyPickerPreference.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    //
+    // Finally, try putting the key inside security hardware, serializing it if that fails.
```
Comment (zopieux) [inline app/src/main/java/androidx/preference/PrivateKeyPickerPreference.kt:38]: "secure hardware"

I've noticed this typo is already throughout existing code, even though Android API confirms the right phrase is *secure* hardware. Either fix it everywhere in this commit, or ignore this comment.

File: build.gradle.kts [hunk 3]
```diff
@@ -27,14 +27,7 @@ buildscript {
...
     dependencies {
-        // R8 below 3.1.79, 3.2.70, and 3.3.62, has an issue;
-        // it tries to convert some Enums to ints but kind of stops halfway.
-        // This was previously solved by pinning R8 to 3.3.62,
-        // but that requires Google's own R8 repo as this version wasn't on Maven,
-        // which is an issue since F-Droid decided to not allowlist that repo.
-        // Instead, we now use Android Gradle Plugin 7.3.0-beta05, which has a fix.
-        // https://issuetracker.google.com/issues/236618700
-        classpath("com.android.tools.build:gradle:7.3.0-beta05")
+        classpath("com.android.tools.build:gradle:7.3.1")
```
Comment (zopieux) [inline build.gradle.kts:30]: Should this be committed separately maybe?

**Review by zopieux — APPROVED** [review]

File: build.gradle.kts (hunk 3 above)
Comment (oakkitten) [reply build.gradle.kts:30]: But it is though

Comment (zopieux) [reply build.gradle.kts:30]: I meant outside the PR, but whatever.

Comment (oakkitten) [reply build.gradle.kts:30]: Oh I now notice your other PR, sure

**Pull request merged**